ruby/ql/lib/codeql/ruby/frameworks/data/internal/ApiGraphModelsSpecific.qll

```diff
@@ -0,0 +1,98 @@
+ * API::Node getExtraSuccessorFromNode(API::Node node, AccessPathToken token)
+ * API::Node getExtraSuccessorFromInvoke(InvokeNode node, AccessPathToken token)
+ * predicate invocationMatchesExtraCallSiteFilter(InvokeNode invoke, AccessPathToken token)
+ * InvokeNode getAnInvocationOf(API::Node node)
+ * ```
+ */
+
+private import ruby
+private import codeql.ruby.DataFlow
+private import codeql.ruby.dataflow.internal.DataFlowPrivate as DataFlowPrivate
+private import ApiGraphModels
+
+class Unit = DataFlowPrivate::Unit;
+
+import codeql.ruby.ApiGraphs
+import codeql.ruby.dataflow.internal.AccessPathSyntax as AccessPathSyntax
+private import AccessPathSyntax
+
+/**
+ * Holds if models describing `package` may be relevant for the analysis of this database.
+ *
+ * In the context of Ruby, this is the name of a Ruby gem.
+ */
+bindingset[package]
+predicate isPackageUsed(string package) {
+  // For now everything is modelled as an access path starting at any top-level, so the package name has no effect.
+  //
+  // We allow an arbitrary package name so that the model can record the name of the package in case it's needed in the future.
+  //
+  // In principle we should consider a package to be "used" if there is a transitive dependency on it, but we can only
+  // reliably see the direct dependencies.
+  //
+  // In practice, packages try to use unique top-level module names, which mitigates the precision loss of not checking
+  // the package name.
+  any()
+}
+
+/** Gets a Ruby-specific interpretation of the `(package, type, path)` tuple after resolving the first `n` access path tokens. */
+bindingset[package, type, path]
+API::Node getExtraNodeFromPath(string package, string type, AccessPath path, int n) {
+  isRelevantFullPath(package, type, path) and
+  package = any(string s) and // Allow any package name, see `isPackageUsed`.
+  type = "" and
+  n = 0 and
+  result = API::root()
+}
+
+/**
+ * Gets a Ruby-specific API graph successor of `node` reachable by resolving `token`.
+ */
+bindingset[token]
+API::Node getExtraSuccessorFromNode(API::Node node, AccessPathToken token) {
+  token.getName() = "Member" and
+  result = node.getMember(token.getAnArgument())
+  or
+  token.getName() = "Method" and
+  result = node.getMethod(token.getAnArgument())
+  or
+  token.getName() = "Instance" and
+  result = node.getInstance()
+  // Note: The "ArrayElement" token is not implemented yet, as it ultimately requires type-tracking and
+  // API graphs to be aware of the steps involving ArrayElement contributed by the standard library model.
+  // Type-tracking cannot summarize function calls on its own, so it doesn't benefit from synthesized callables.
+}
+
+/**
+ * Gets a Ruby-specific API graph successor of `node` reachable by resolving `token`.
+ */
+bindingset[token]
+API::Node getExtraSuccessorFromInvoke(InvokeNode node, AccessPathToken token) {
+  token.getName() = "Instance" and
+  result = node.getInstance()
+}
```

Comment:
I'm not sure what we are trying to match here. `InvokeNode` corresponds to a method call, which doesn't have a concept of an instance, as far as I know. The closest thing would be if it returns a class which can be instantiated e.g. `foo().new`, but that should be handled in `getExtraSuccessorFromNode` since the path would be `Method[foo].ReturnValue.Instance`.

Is this maybe a JS-ism that doesn't translate to Ruby, like `new foo()` or something?